### ISSUE TITLE
API Gateway Doesn't Purge "Lost" Records

### ISSUE THREAD

**Issue by plombardi89** [datawire/gateway#1]
If the Discovery listener drops a Discovery message for whatever reason then it's possible for the API Gateway to fall out of sync with the rest of the system. This can have disastrous consequences as the underlying Traefik server continues to attempt to route messages to the lost backends.

A possible solution is to periodically disconnect and reconnect thus forcing a fresh state transfer from the Discovery backend.

Another option is for the Discovery server to periodically blast the state of the world to all clients.

A third option is to provide a query interface (similar to the reconnect option).

**Comment by rhs** [comment]
Shouldn't the API gateway also have some kind of circuit breaker behavior? I'm thinking you could get a similar situation even without losing a record if the node dies but the record is still in the routing table.

**Comment by plombardi89** [comment]
Traefik has some built in capabilities for circuit breaking. We don't tap into them right now. You're right we should probably use them, but having a way to sync with the Discovery system is also important IMO.

**Comment by rhs** [comment]
Agreed, I just wanted to point out that strictly speaking I think there are probably two distinct bugs here.

**Comment by plombardi89** [comment]
The problem with the Traefik CB mechanism is that it's a rolling window which eventually clears. If you have a server that can never recover (for example someone sent a null/None address) then it will be stuck forever and Traefik will continue to try and serve it after the tripped timeout clears.

https://docs.traefik.io/basics/#backends

But yea definitely two issues.
